### ISSUE TITLE
Figure out if we can support java streams (and if not, why not)

### ISSUE THREAD

**Issue by gzm0** [scala-js/scala-js#4926]
Following up from
https://github.com/scala-js/scala-js/issues/4925#issuecomment-1878517715_

IMHO Java streams are sufficiently important that if we cannot support them, we should at the very least formally document that (and why).

**Comment by gzm0** [comment]
I've had a look at the `java.util.stream` package and I really do not see any major reason why we cannot support it.

Of course, we will not be able to offer parallel streams. This means, we should probably not even define all the parallel methods.

There is the (unfortunate) nuisance that the methods on [`StreamSupport`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/StreamSupport.html) take `parallel` as an argument. So we'd have to resort to throwing `UnsupportedOperationException` in this case (which given how low-level the class is, seems acceptable to me).

@sjrd, what am I missing? :)

**Comment by gzm0** [comment]
Note on [Collection#parallelStream](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#parallelStream()).

> Returns a possibly parallel Stream with this collection as its source. It is allowable for this method to return a sequential stream.

So just ignoring request for parallelism might be just fine.

**Comment by sjrd** [comment]
I had another look and I cannot find anything that is a total deal-breaker. The `StreamSupport` issue is annoying, but perhaps we can simply *not* provide `StreamSupport` at all. It is a static-only helper class, which is not interleaved in the rest of the API.

The fact that `parallel()` is abstract in `BaseStream` and its public subinterfaces means that third-party code will have trouble implementing their own stream subclasses: if they don't declare an implementation of `parallel()`, the compiler will complain that it is left abstract; but if they do, they inevitably break its contract. Some IR post-processing may fix that issue, but that's harsh to ask of third-party implementers.

> Note on [Collection#parallelStream](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#parallelStream()).
> 
> > Returns a possibly parallel Stream with this collection as its source. It is allowable for this method to return a sequential stream.
> 
> So just ignoring request for parallelism might be just fine.

It does have an **Implementation Requirements** section that says that the default implementation returns a parallel stream, so I think it is still off-limits.

**Comment by gzm0** [comment]
>  The StreamSupport issue is annoying, but perhaps we can simply not provide StreamSupport at all. It is a static-only helper class, which is not interleaved in the rest of the API.

Fair enough. Feels to me like we definitely want something like this to implement streams, but at least for starters, it can be private to the javalib without the `parallel` parameter.

> Some IR post-processing may fix that issue, but that's harsh to ask of third-party implementers.

True. But I guess we agree that providing streams simply for usage is worth it in its own right at least as a first step?

> It does have an Implementation Requirements section that says that the default implementation returns a parallel stream, so I think it is still off-limits.

True :-/

In any case, do you agree that based on this, we should re-open #4925?

**Comment by sjrd** [comment]
> > Some IR post-processing may fix that issue, but that's harsh to ask of third-party implementers.
> 
> True. But I guess we agree that providing streams simply for usage is worth it in its own right at least as a first step?

Yes, it's probably viable for usage-only at first.

> In any case, do you agree that based on this, we should re-open #4925?

Reopened.

**Comment by He-Pin** [comment]
As [Stream Gatherers](https://openjdk.org/jeps/485) will be released soon, we may implemented other operators based on that.
